op-node/rollup/derive/channel.go

```diff
@@ -147,8 +151,41 @@ func (ch *Channel) Reader() io.Reader {
 // Warning: the batch reader can read every batch-type.
 // The caller of the batch-reader should filter the results.
 func BatchReader(r io.Reader) (func() (*BatchData, error), error) {
+	// Buffer to store the data for reuse
+	var buffer strings.Builder
+	if _, err := io.Copy(&buffer, r); err != nil {
+		return nil, err
+	}
+
+	tmpReader := strings.NewReader(buffer.String())
+	// Read the first byte to be used to determine the compression type
+	compressionType := make([]byte, 1)
+	if _, err := tmpReader.Read(compressionType); err != nil {
+		return nil, err
+	}
+
+	// Reset the reader with the original data
+	r = strings.NewReader(buffer.String())
```

Comment:
Had to do this because reading the original buffer will increment the read cursor, but zlib needs to read from the beginning.
Curious if there's a better way that doesn't require this

Reply:
You could wrap it in a buffered reader and use https://pkg.go.dev/bufio#Reader.Peek ?